sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2109,6 +2110,51 @@ class Analyzer(override val catalogManager: CatalogManager)
     }
   }
 
+  /**
+   * Resolves `UnresolvedAttribute` to `OuterReference` if we are resolving subquery plans (when
+   * `AnalysisContext.get.outerPlan` is set).
+   */
+  object ResolveOuterReferences extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = {
+      // Only apply this rule if we are resolving subquery plans.
+      if (AnalysisContext.get.outerPlan.isEmpty) return plan
+
+      // We must run these 3 rules first, as they also resolve `UnresolvedAttribute` and have
+      // higher priority than outer reference resolution.
+      val prepared = ResolveAggregateFunctions(ResolveMissingReferences(ResolveReferences(plan)))
```

Comment:
Aren't `ResolveMissingReferences` and `ResolveReferences` already run before `ResolveOuterReferences`?

Reply:
The key is to run `ResolveOuterReferences` **right after** these 3 rules, so that it's safe to resolve `UnresolvedAttribute` to outer references. Otherwise, other rules may change the plan shape and make these 3 rules applicable again.

Reply:
I guess one disadvantage of running these three rules inside ResolveOuterReferences is that they are not visible in the plan change log.

Reply:
Also it's a bit strange when there is no outer reference to resolve but the 3 rules take effect, the plan change logger of this ResolveOuterReferences actually shows the changes from these 3 rules.
```
=== Applying Rule org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveOuterReferences ===
 'Project [id2#28L]                                                                                          'Project [id2#28L]
 +- 'SubqueryAlias __auto_generated_subquery_name                                                            +- 'SubqueryAlias __auto_generated_subquery_name
!   +- 'Project [id#27, ('id + cast(1 as bigint)) AS id2#28]                                                    +- 'Project [id#27, (id#27 + cast(1 as bigint)) AS id2#28]
..
```

Reply:
Yes, this is not a perfect solution, but AFAIK this is the only reliable way to guarantee rule execution order. The best solution in my opinion is to centralize all column resolution code in one rule, but that's a much larger change.